File: backend/app/models/giveaway.py
"""Giveaway models (FEATURES_COMMUNITY_PLUS · Giveaway).

Peer-to-peer free clothes. Listings are public; contact happens in-app via a
private claim message — no personal address/phone in public listings (§10).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class GiveawayCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    images: list[str] = Field(default_factory=list)  # public image URLs
    size: str | None = Field(default=None, max_length=60)
    category: str | None = Field(default=None, max_length=60)
    condition: str | None = Field(default=None, max_length=60)
    area_label: str | None = Field(default=None, max_length=120)
    wardrobe_item_id: UUID | None = None

    @field_validator("images")
    @classmethod
    def _clean_images(cls, value: list[str]) -> list[str]:
        return [u.strip() for u in value if u.strip()][:6]


class GiveawayResponse(BaseModel):
    id: str
    owner_id: str
    owner_name: str | None = None
    wardrobe_item_id: str | None = None
    title: str
    description: str | None = None
    images: list[str] = Field(default_factory=list)  # full images (R2 CDN or passthrough)
    thumbnails: list[str] = Field(default_factory=list)  # smaller, parallel to images
    size: str | None = None
    category: str | None = None
    condition: str | None = None
    area_label: str | None = None
    status: str
    is_mine: bool = False
    my_claim_status: str | None = None  # the caller's own claim status, if any
    claim_count: int = 0
    created_at: datetime


class ClaimCreate(BaseModel):
    message: str | None = Field(default=None, max_length=1000)


class ClaimResponse(BaseModel):
    id: str
    giveaway_id: str
    claimer_id: str
    claimer_name: str | None = None
    message: str | None = None
    status: str
    created_at: datetime


class GiveawayStatusUpdate(BaseModel):
    status: Literal["available", "reserved", "claimed", "closed"]


class ClaimDecision(BaseModel):
    status: Literal["accepted", "declined"]


# ── Secret pickup chat (0037) ────────────────────────────────────────────────
# Private owner ↔ accepted-requester coordination, active for 7 days from the
# accept. Text-only, ≤500 chars; bodies are redacted after the chat ends (§10).


class ChatMessageCreate(BaseModel):
    body: str = Field(min_length=1, max_length=500)

    @field_validator("body")
    @classmethod
    def _strip_body(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("message can't be empty")
        return value


class ChatMessageResponse(BaseModel):
    id: str
    chat_id: str
    sender_id: str
    is_mine: bool = False
    body: str | None = None  # None once redacted
    body_deleted: bool = False
    created_at: datetime


class PickupPlanUpdate(BaseModel):
    """The pickup plan card — coarse, public-place info ONLY (§10)."""

    area: str | None = Field(default=None, max_length=120)  # suburb / general area
    landmark: str | None = Field(default=None, max_length=160)  # public point
    time_slot: str | None = Field(default=None, max_length=120)
    confirmed: bool = False


class ChatReportCreate(BaseModel):
    reason: str | None = Field(default=None, max_length=500)


class PickupChatResponse(BaseModel):
    id: str
    giveaway_id: str
    giveaway_title: str | None = None
    owner_id: str
    requester_id: str
    other_name: str | None = None  # display name of the OTHER participant
    is_owner: bool = False  # whether the caller is the giveaway owner
    status: str  # active | completed | cancelled | expired (locked lifecycle)
    report_flag: bool = False
    pickup_plan: dict = Field(default_factory=dict)
    approved_at: datetime
    expires_at: datetime
    locked_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime

File: app/lib/data/models/giveaway.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'giveaway.freezed.dart';
part 'giveaway.g.dart';

/// A claim on a giveaway (FEATURES_COMMUNITY_PLUS · Giveaway). The [message] is
/// private to the giveaway owner (in-app contact only).
@freezed
abstract class GiveawayClaim with _$GiveawayClaim {
  const factory GiveawayClaim({
    required String id,
    @JsonKey(name: 'giveaway_id') required String giveawayId,
    @JsonKey(name: 'claimer_id') required String claimerId,
    @JsonKey(name: 'claimer_name') String? claimerName,
    String? message,
    // requested | accepted | declined | not_selected | cancelled | expired
    required String status,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _GiveawayClaim;

  factory GiveawayClaim.fromJson(Map<String, dynamic> json) =>
      _$GiveawayClaimFromJson(json);
}

/// The secret pickup chat between the giveaway owner and the ONE accepted
/// requester — active for exactly 7 days from the accept, then locked and
/// redacted server-side. Never visible to anyone else (§10).
@freezed
abstract class GiveawayPickupChat with _$GiveawayPickupChat {
  const factory GiveawayPickupChat({
    required String id,
    @JsonKey(name: 'giveaway_id') required String giveawayId,
    @JsonKey(name: 'giveaway_title') String? giveawayTitle,
    @JsonKey(name: 'owner_id') required String ownerId,
    @JsonKey(name: 'requester_id') required String requesterId,
    /// Display name of the OTHER participant (server picks per caller).
    @JsonKey(name: 'other_name') String? otherName,
    @JsonKey(name: 'is_owner') @Default(false) bool isOwner,
    // active | locked | completed | cancelled | expired
    required String status,
    @JsonKey(name: 'report_flag') @Default(false) bool reportFlag,
    @JsonKey(name: 'pickup_plan')
    @Default(<String, dynamic>{})
    Map<String, dynamic> pickupPlan,
    @JsonKey(name: 'approved_at') required DateTime approvedAt,
    @JsonKey(name: 'expires_at') required DateTime expiresAt,
    @JsonKey(name: 'locked_at') DateTime? lockedAt,
    @JsonKey(name: 'completed_at') DateTime? completedAt,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _GiveawayPickupChat;

  const GiveawayPickupChat._();

  factory GiveawayPickupChat.fromJson(Map<String, dynamic> json) =>
      _$GiveawayPickupChatFromJson(json);

  /// Open for messages: still `active` AND inside the 7-day window (the app
  /// double-checks the clock so a stale fetch can't look sendable).
  bool get isActive =>
      status == 'active' && DateTime.now().toUtc().isBefore(expiresAt.toUtc());

  Duration get timeLeft {
    final left = expiresAt.toUtc().difference(DateTime.now().toUtc());
    return left.isNegative ? Duration.zero : left;
  }

  String? get planArea => pickupPlan['area'] as String?;
  String? get planLandmark => pickupPlan['landmark'] as String?;
  String? get planTimeSlot => pickupPlan['time_slot'] as String?;
  bool get planConfirmed => pickupPlan['confirmed'] == true;
  bool get hasPlan =>
      (planArea ?? planLandmark ?? planTimeSlot) != null || planConfirmed;
}

/// One text message in a pickup chat. [body] is null once the retention job
/// has redacted it ([bodyDeleted]).
@freezed
abstract class GiveawayChatMessage with _$GiveawayChatMessage {
  const factory GiveawayChatMessage({
    required String id,
    @JsonKey(name: 'chat_id') required String chatId,
    @JsonKey(name: 'sender_id') required String senderId,
    @JsonKey(name: 'is_mine') @Default(false) bool isMine,
    String? body,
    @JsonKey(name: 'body_deleted') @Default(false) bool bodyDeleted,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _GiveawayChatMessage;

  factory GiveawayChatMessage.fromJson(Map<String, dynamic> json) =>
      _$GiveawayChatMessageFromJson(json);
}

/// A peer-to-peer free-clothes listing. Public listing fields only — contact is
/// arranged in-app via a claim, never personal address/phone in the listing.
@freezed
abstract class Giveaway with _$Giveaway {
  const factory Giveaway({
    required String id,
    @JsonKey(name: 'owner_id') required String ownerId,
    @JsonKey(name: 'owner_name') String? ownerName,
    @JsonKey(name: 'wardrobe_item_id') String? wardrobeItemId,
    required String title,
    String? description,
    @Default(<String>[]) List<String> images,
    // Smaller images parallel to [images] (grid cover), where available.
    @Default(<String>[]) List<String> thumbnails,
    String? size,
    String? category,
    String? condition,
    @JsonKey(name: 'area_label') String? areaLabel,
    required String status, // available | reserved | claimed | closed
    @JsonKey(name: 'is_mine') @Default(false) bool isMine,
    @JsonKey(name: 'my_claim_status') String? myClaimStatus,
    @JsonKey(name: 'claim_count') @Default(0) int claimCount,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _Giveaway;

  const Giveaway._();

  factory Giveaway.fromJson(Map<String, dynamic> json) =>
      _$GiveawayFromJson(json);

  bool get isAvailable => status == 'available';
  bool get hasClaimed => myClaimStatus != null;

  /// The grid cover image: the first thumbnail where available, else the first
  /// full image, else null.
  String? get coverImageUrl => thumbnails.isNotEmpty
      ? thumbnails.first
      : (images.isNotEmpty ? images.first : null);
}
